commonmark/src/main/java/org/commonmark/internal/ReaderParser.java

```diff
@@ -0,0 +1,24 @@
+package org.commonmark.internal;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.Reader;
+import java.util.List;
+import org.commonmark.parser.block.BlockParserFactory;
+
+public class ReaderParser extends DocumentParser<Reader> {
+
+    public ReaderParser(List<BlockParserFactory> blockParserFactories, InlineParserImpl inlineParser) {
+        super(blockParserFactories, inlineParser);
+    }
+    
+    @Override
+    protected void incorporateLines(Reader input) throws IOException {
+        try (BufferedReader r = new BufferedReader(input)){
```

Comment:
This causes the passed `input` to be closed at the end of the block, which is not what we want. We want the caller to handle this, we will just read from it until we're done. The typical caller then looks like this:

```java
try (BufferedReader reader = Files.newBufferedReader(filePath)) {
    Node node = parser.parse(reader);
    /* ... */
}
```

Also, I think we should do the following:

```java
BufferedReader bufferedReader;
if (input instanceof BufferedReader) {
    bufferedReader = (BufferedReader) input;
} else {
    bufferedReader = new BufferedReader(input);
}
```